### PR TITLE
Add simplification rule for bitwise shifting

### PR DESCRIPTION
**Pull request by axic**
Depends on #2541.

### CODE REVIEW

File: libevmasm/RuleList.h [hunk 1]
```diff
@@ -89,6 +89,16 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleList(
 			u256 mask = (u256(1) << testBit) - 1;
 			return u256(boost::multiprecision::bit_test(B.d(), testBit) ? B.d() | ~mask : B.d() & mask);
 		}, false},
+		{{Instruction::SHL, {A, B}}, [=]{
```
Comment (axic) [inline libevmasm/RuleList.h:92]: @chriseth do we want to introduce `evmVersion` in the rule list or is it safe to have this merged (after some cleanup/squashing)?

Comment (chriseth) [reply libevmasm/RuleList.h:92]: I think it should be safe. Let's add a comment.

Comment (chriseth) [reply libevmasm/RuleList.h:92]: Are you sure about the order of the operands?

Comment (chriseth) [reply libevmasm/RuleList.h:92]: Please add a test.

Comment (chriseth) [reply libevmasm/RuleList.h:92]: (is that possible?)

Comment (axic) [reply libevmasm/RuleList.h:92]: This is blocked by the same problem as in #3797: need to build a constantinople supporting version of `eth` and run endtoend tests with that target.

**Comment by axic** [comment]
@pirapira can you check this, since you have written tests for shifts?

File: libevmasm/RuleList.h [hunk 2]
```diff
@@ -89,6 +89,16 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleList(
...
+			if (A.d() > 255)
+				return u256(0);
+			return u256(bigint(B.d()) << unsigned(A.d()));
+		}},
+		{{Instruction::SHR, {A, B}}, [=]{
+			if (A.d() > 255)
+				return u256(0);
+			return B.d() >> unsigned(A.d());
```
Comment (pirapira) [inline libevmasm/RuleList.h]: Is `B.d()` signed or unsigned?  If signed, can it be negative?

Comment (chriseth) [reply libevmasm/RuleList.h]: It always returns `u256`.

File: libevmasm/RuleList.h (hunk 1 above)
Comment (axic) [reply libevmasm/RuleList.h:92]: > Are you sure about the order of the operands?

The top of the stack is the number of bits to shift he value with.

File: libevmasm/RuleList.h (hunk 2 above)
Comment (axic) [reply libevmasm/RuleList.h]: `SHR` always considers `B` as unsigned. There is no optimisation rule adde for `SAR` as I was lazy.

**Comment by axic** [comment]
@chfast @pirapira do you want to double check the expectations in the endtoend tests?

@chriseth this should be ready to merge

**Review by chfast — APPROVED** [review]
For constants this is ok.
What was the other one what tried to optimize division?

**Review by chriseth — APPROVED** [review]

**Pull request merged**